internal/pkg/gateway/commit/transactionstatus.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2021 IBM All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package commit
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric-protos-go/peer"
+	"github.com/hyperledger/fabric/common/ledger/blockledger"
+	"github.com/pkg/errors"
+)
+
+// TransactionStatus is sent to the returned channel when received from the block iterator. The returned channel will
+// be closed after the status is sent, or if an error occurs that prevents the transaction status from being obtained.
+// The receiver should always check the the closed status of the channel on reads.
+func TransactionStatus(blockIterator blockledger.Iterator, transactionID string, quit <-chan int) <-chan peer.TxValidationCode {
+	out := make(chan peer.TxValidationCode, 1)
+	go func() {
+		defer close(out)
+
+		for {
+			select {
+			case <-quit:
+				return
+			default:
+			}
+
+			block, status := blockIterator.Next()
+			if status != common.Status_SUCCESS {
+				// TODO: log error -- ledger is broken
+				return
+			}
+
+			if validationCode, found := findTransactionStatus(block, transactionID); found {
+				out <- validationCode
+				return
+			}
+		}
+	}()
+
+	return out
+}
+
+func findTransactionStatus(block *common.Block, transactionID string) (peer.TxValidationCode, bool) {
+	envelopes := block.Data.Data
```

Comment:
Does this need nil checks for blocks committed to the peer ledger?

Reply:
Internally in the peer, the nil check makes little sense.  If you want this code to be portable outside the peer, then it would be a good idea.  My feeling is that we omit the check for now.